src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -331,18 +553,441 @@ namespace Opm {
+                coord.push_back(x0);
+                coord.push_back(y0);
+                coord.push_back(zb);
+
+                for (int i = 0; i < dims[0]; i++) {
+
+                    size_t ind = i+j*dims[0]+1;
+
+                    if (i == (dims[0]-1)) {
+                        ind=ind-1;
+                    }
+
+                    zt = tops[ind];
+                    zb = zt + sumDZ(i, j, dims, dz);
+
+                    double xt = x0 + dx[i+j*dims[0]];
+                    double xb = sumDX(j, dims[2]-1, i, dims, dx);
+
+                    coord.push_back(xt);
+                    coord.push_back(y0);
+                    coord.push_back(zt);
+                    coord.push_back(xb);
+                    coord.push_back(y0);
+                    coord.push_back(zb);
+
+                    x0=xt;
+                }
+            }
+
+            int ind = (j+1)*dims[0];
+
+            if (j == (dims[1]-1) ) {
+                ind = j*dims[0];
+            }
+
+            double x0 = 0.0;
+
+            double yt = sumDY(0, 0, j, dims, dy);
+            double yb = sumDY(0, dims[2]-1, j, dims, dy);
+
+            zt = tops[ind];
+            zb = zt + sumDZ(0, j, dims, dz);
+
+            coord.push_back(x0);
+            coord.push_back(yt);
+            coord.push_back(zt);
+            coord.push_back(x0);
+            coord.push_back(yb);
+            coord.push_back(zb);
+
+            for (int i=0; i < dims[0]; i++) {
+
+                ind = i+(j+1)*dims[0]+1;
+
+                if (j == (dims[1]-1) ) {
+                    ind = i+j*dims[0]+1;
+                }
+
+                if (i == (dims[0] - 1) ) {
+                    ind=ind-1;
+                }
+
+                zt = tops[ind];
+                zb = zt + sumDZ(i, j, dims, dz);
+
+                double xt=-999;
+                double xb=-999;
+
+                if (j == (dims[1]-1) ) {
+                    xt = sumDX(j, 0, i, dims, dx);
+                    xb = sumDX(j, dims[2]-1, i, dims, dx);
+                } else {
+                    xt = sumDX(j+1, 0, i, dims, dx);
+                    xb = sumDX(j+1, dims[2]-1, i, dims, dx);
+                }
+
+                if (i == (dims[0] - 1) ) {
+                    yt = sumDY(i, 0, j, dims, dy);
+                    yb = sumDY(i, dims[2]-1, j, dims, dy);
+                } else {
+                    yt = sumDY(i+1, 0, j, dims, dy);
+                    yb = sumDY(i+1, dims[2]-1, j, dims, dy);
+                }
+
+                coord.push_back(xt);
+                coord.push_back(yt);
+                coord.push_back(zt);
+                coord.push_back(xb);
+                coord.push_back(yb);
+                coord.push_back(zb);
+
+                x0 = xt;
+            }
+        }
+
+        return coord;
+    };
+
+    std::vector<double> EclipseGrid::makeZcornDzTops(const std::array<int, 3>& dims, const std::vector<double>& dz, const std::vector<double>& tops) const {
```

Comment:
I have not managed to read it thoroughly - but to me it looks like it should be possible to extract common functionality from the various makeZcornxxx` and `makeCoordxxx` functions?